### PR TITLE
DEP: signal: raise error using medfilt and order_filter with float128 and object dtypes

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

follow up to #18341
#### What does this implement/fix?

It has been two releases since float128 and object types were deprecated to deduplicate code, should be fine to raise an error now.

### CODE REVIEW

File: scipy/signal/tests/test_signaltools.py [hunk 1]
```diff
@@ -1115,7 +1106,7 @@ def test_invalid_dtypes(self, dtype):
     def test_none(self):
         # gh-1651, trac #1124. Ensure this does not segfault.
         with pytest.warns(UserWarning):
-            assert_raises(TypeError, signal.medfilt, None)
+            assert_raises(ValueError, signal.medfilt, None)
```
Comment (j-bowhay) [inline scipy/signal/tests/test_signaltools.py:1109]: Looks like there was some sort of special casing here for None. Have it lumped in with the others that raise ValueError but could special case so this remains a TypeError

File: scipy/signal/tests/test_signaltools.py [hunk 2]
```diff
@@ -1124,30 +1115,6 @@ def test_none(self):
         a.strides = 16
         assert_(signal.medfilt(a, 1) == 5.)
 
-    def test_refcounting(self):
```
Comment (j-bowhay) [inline scipy/signal/tests/test_signaltools.py:1127]: Object array test hopefully safe to remove

File: scipy/signal/_signaltools.py
Comment (h-vetinari) [inline scipy/signal/_signaltools.py:1]: Changes in this file are fine

File: scipy/signal/_sigtoolsmodule.c
Comment (h-vetinari) [inline scipy/signal/_sigtoolsmodule.c:1]: But the changes here are extensive and not at all indicated by the PR title.

I'm quite surprised that we can just delete `_sigtools._order_filterND` for _all_ dtypes (not just float128/object), and is that necessary/worthwhile? That's IMO way beyond just executing the deprecation (i.e. what's in `scipy/signal/_signaltools.py`).

**Review by h-vetinari — COMMENTED** [review]
There's a mismatch between the PR title and the scope of what it does.

**Comment by lucascolley** [comment]
I'm also a bit confused since gh-18341 said "The goal here is to reduce duplication between ndimage and signal, and get rid of https://github.com/scipy/scipy/blob/main/scipy/signal/_lfilter.c.in entirely (700 LOC of fairly obscure C code). Cannot just remove them, so deprecate first."

**Comment by j-bowhay** [comment]
The point of the original deprecation was that we currently have two sets of C code for doing the same thing, one in `signal` and one in `ndimage`. The signal functions were modified to use the ndimage code under the hood so signal C code could be removed as it was decided the `ndimage` code was more maintainable. However `ndimage` did not support object or float128 dtypes hence the deprecation cycle before all the `signal` C code could be removed.

**Comment by j-bowhay** [comment]
There are some relevant test failures that need resolving though

**Comment by lucascolley** [comment]
Ah okay, so does all of `_lfilter.c.in` need to stay?

**Comment by j-bowhay** [comment]
> Ah okay, so does all of `_lfilter.c.in` need to stay?

As far as I could tell it was still needed but I also might have just got lost the maze. Maybe @ev-br could confirm?

File: scipy/signal/tests/test_signaltools.py (hunk 2 above)
Comment (rgommers) [reply scipy/signal/tests/test_signaltools.py:1127]: This should be fine - and we should remove all tests that explicitly check the refcount (I'm not sure if we have any others), because with changes in CPython those tests can no longer be correct.

File: scipy/signal/_signaltools.py [hunk 3]
```diff
@@ -1565,25 +1560,13 @@ def medfilt(volume, kernel_size=None):
...
                       stacklevel=2)
 
-    domain = np.ones(kernel_size, dtype=volume.dtype)
-
-    numels = np.prod(kernel_size, axis=0)
-    order = numels // 2
-
     if volume.dtype in [np.bool_, np.complex64, np.complex128, np.clongdouble,
-                        np.float16]:
+                        np.float16, np.object_, 'float128']:
```
Comment (rgommers) [inline scipy/signal/_signaltools.py:1564]: Note that you want to use `longdouble` as the dtype name, because that's the canonical one (`float128` may not exist, and then the alias `float96` will exist)

**Comment by ev-br** [comment]
> Ah okay, so does all of _lfilter.c.in need to stay?

Indeed. Thank you Jake for chasing down where this dead code is despite the typo in the original PR which introduced the deprecation.

**Comment by j-bowhay** [comment]
Updated to use `longdouble` and removed more dead code

File: scipy/signal/tests/test_signaltools.py (hunk 1 above)
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1109]: I dislike the Warning+Error combination, aside from the use of `assert_raises`. It would be good to just raise on None directly, without warning spuriously.

Looking at the code, this is because `volume=None, kernel_size=None` ends up being transformed to
```
>>> volume
array([None], dtype=object)
>>> kernel_size
[3]
```
before being fed into

https://github.com/scipy/scipy/blob/bc41135d2dc01b5d99ac2b0a1ae28951fa42f9e5/scipy/signal/_signaltools.py#L1563-L1566

IMO, the test should roughly look as follows
```python
    def test_none(self):
        # gh-1651, trac #1124. Ensure this does not segfault.
        with pytest.raises(ValueError, match="..."):
            signal.medfilt(None)
```
This will require a small change in medfilt, probably easiest to add a branch `if volume is None: raise ...`.

File: scipy/signal/tests/test_signaltools.py [hunk 4]
```diff
@@ -1115,7 +1106,7 @@ def test_invalid_dtypes(self, dtype):
     def test_none(self):
         # gh-1651, trac #1124. Ensure this does not segfault.
         with pytest.warns(UserWarning):
-            assert_raises(TypeError, signal.medfilt, None)
+            assert_raises(ValueError, signal.medfilt, None)
         # Expand on this test to avoid a regression with possible contiguous
         # numpy arrays that have odd strides. The stride value below gets
         # us into wrong memory if used (but it does not need to be used)
```
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1112]: This is not the fault of this PR (actually, since the introduction of this hunk in https://github.com/scipy/scipy/commit/ce7e474b3df009c36a9710276d98875eecb23b2c), but this second part of the test has nothing to do with `test_none`, and should just get a separate `def test_odd_strides` (or whatever other name).

**Review by h-vetinari — COMMENTED** [review]
Some relatively minor nits, otherwise looking good (thanks for the context on `_order_filterND`).

File: scipy/signal/_signaltools.py [hunk 5]
```diff
@@ -1551,6 +1546,10 @@ def medfilt(volume, kernel_size=None):
 
     """
     volume = np.atleast_1d(volume)
+    if volume.dtype in [np.bool_, np.complex64, np.complex128, np.clongdouble,
+                        np.float16, np.object_, np.longdouble]:
+        raise ValueError(f"dtype={volume.dtype} is not supported by medfilt")
```
Comment (h-vetinari) [inline scipy/signal/_signaltools.py:1551]: I think `np.longdouble` here is incorrect, because on windows that just ends up being `float64` and then causes
```
FAILED scipy\signal\tests\test_signaltools.py::TestMedFilt::test_types[float64] - ValueError: dtype=float64 is not supported by medfilt
```

IMO it would be better to formulate this as:
```suggestion
    if volume.dtype not in <list_of_allowed_types>:
        raise ValueError(f"dtype={volume.dtype} is not supported by medfilt")
```

**Review by h-vetinari — COMMENTED** [review]
Last details. Not sure why the doc build is failing, but at least win/macos jobs need to be fixed.

File: scipy/signal/_signaltools.py [hunk 6]
```diff
@@ -1551,6 +1546,10 @@ def medfilt(volume, kernel_size=None):
 
     """
     volume = np.atleast_1d(volume)
+    if volume.dtype in [np.bool_, np.complex64, np.complex128, np.clongdouble,
+                        np.float16, np.object_, "float128", "float96"]:
+        raise ValueError(f"dtype={volume.dtype} is not supported by medfilt")
```
Comment (h-vetinari) [inline scipy/signal/_signaltools.py:1551]: I would still prefer explicitly naming the types for which we have support in medfilt (non-binding).

File: scipy/signal/tests/test_signaltools.py [hunk 7]
```diff
@@ -1093,20 +1093,13 @@ def test_types(self, dtype):
...
-
-
     @pytest.mark.parametrize('dtype', [np.bool_, np.complex64, np.complex128,
-                                       np.clongdouble, np.float16,])
+                                       np.clongdouble, np.float16, np.longdouble,
+                                       np.object_])
     def test_invalid_dtypes(self, dtype):
         in_typed = np.array(self.IN, dtype=dtype)
+        if in_typed.dtype is np.float64:
+            pytest.mark.skip("Platform does not support `float128`")
```
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1102]: The combination of if-condition and skip-message makes no sense. It would be much better not to parametrize on an ambiguous type (`np.longdouble`) in the first place, but if that's unavoidable, one way to clarify would be (feel free to rewrite):
```suggestion
        if in_typed.dtype is np.float64:
            pytest.mark.skip("Platform treats `np.longdouble` as `np.float64`; "
                             "we support the latter, but not the former (in general).")
```

**Review by h-vetinari — COMMENTED** [review]
Almost there, but still not yet. 🙃

File: scipy/signal/_signaltools.py (hunk 6 above)
Comment (j-bowhay) [reply scipy/signal/_signaltools.py:1551]: If starting from scratch I would entirely agree with you however I am slightly hesitant to do this as, given all the longdouble confusion, I think it might be quite easy to mess up and introduce a change that was not within the scope of the original deprecation.

File: scipy/signal/tests/test_signaltools.py (hunk 7 above)
Comment (j-bowhay) [reply scipy/signal/tests/test_signaltools.py:1102]: Agh, the tests are still failing on windows and macos arm, neither of which I have access to so this clearly is not quite right. @ev-br would you be able to point me in the right direction here as I am getting quite confused!

Comment (ev-br) [reply scipy/signal/tests/test_signaltools.py:1102]: I don't think `if dtype is np.float64` is correct really. I suspect you're hitting https://github.com/scipy/scipy/pull/18341#issuecomment-1532754157, so can dispatch on `in_typed.dtype.char == 'g'` instead?

I can check it in a couple of days unless you beat me to it.

File: scipy/signal/_signaltools.py (hunk 6 above)
Comment (h-vetinari) [reply scipy/signal/_signaltools.py:1551]: My point is that it should be trivial to specify the types we support - those are the one we actually implement (resp. template over), and that's IMO easier and more useful than playing whack-a-mole with weird/rare types that we somehow forgot to mention here.

File: scipy/signal/tests/test_signaltools.py (hunk 7 above)
Comment (h-vetinari) [reply scipy/signal/tests/test_signaltools.py:1102]: Or just don't parametrize the test with `np.longdouble`, but rather `float96` & `float128`.

Comment (j-bowhay) [reply scipy/signal/tests/test_signaltools.py:1102]: Maybe I am misunderstanding but I don't think we can use `float128` directly here as it doesn't exist on some platforms.

File: scipy/signal/tests/test_signaltools.py [hunk 8]
```diff
@@ -1093,20 +1093,13 @@ def test_types(self, dtype):
...
-        with pytest.deprecated_call(match=msg):
-            assert_equal(signal.medfilt2d(in_typed).dtype, dtype)
-
-
     @pytest.mark.parametrize('dtype', [np.bool_, np.complex64, np.complex128,
-                                       np.clongdouble, np.float16,])
+                                       np.clongdouble, np.float16, np.longdouble,
+                                       np.object_])
     def test_invalid_dtypes(self, dtype):
         in_typed = np.array(self.IN, dtype=dtype)
```
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1100]:
```suggestion
                                       np.clongdouble, np.float16,
                                       np.object_, "float96", "float128"])
    def test_invalid_dtypes(self, dtype):
        if system_does_not_have_large_float and dtype in ["float96", "float128"]:
            pytest.mark.skip("Platform does not support `float128`")
        in_typed = np.array(self.IN, dtype=dtype)
```

`system_does_not_have_large_float` is left as an exercise to the reader. ;-)

File: scipy/signal/tests/test_signaltools.py (hunk 7 above)
Comment (h-vetinari) [reply scipy/signal/tests/test_signaltools.py:1102]: Yes, you'll need to skip it on platforms that don't have it. I made another suggestion that's IMO easier to follow, because there's no need to invert any logic (of the kind "we support everything but these types" or "we support `np.longdouble` except when it's actually `np.float64`).

**Comment by h-vetinari** [comment]
I think we can wait for 1.14 with this? Unless you want to get this in urgently, I think we should give it a year as usual.

**Comment by j-bowhay** [comment]
Closing until we branch for 1.14

**Comment by h-vetinari** [comment]
Branch has happened, reopening.

**Comment by h-vetinari** [comment]
@j-bowhay, do you think you'll be able to rebase this in the coming weeks?

File: scipy/signal/_signaltools.py (hunk 6 above)
Comment (h-vetinari) [reply scipy/signal/_signaltools.py:1551]: Thanks for picking this up again! I just looked through again and it looks good, though I'd still want to allowlist certain types here, rather than blocklist an arbitrarily long list of dtypes (as an example taken at random, we're probably not failing correctly for datetime types)

Comment (j-bowhay) [reply scipy/signal/_signaltools.py]: Thanks for taking a look, hopefully this has been addressed in the latest commit

File: scipy/signal/_signaltools.py [hunk 9]
```diff
@@ -1496,15 +1496,10 @@ def order_filter(a, domain, rank):
...
-    if a.dtype in [object, 'float128']:
-        mesg = (f"Using order_filter with arrays of dtype {a.dtype} is "
-                f"deprecated in SciPy 1.11 and will be removed in SciPy 1.14")
-        warnings.warn(mesg, DeprecationWarning, stacklevel=2)
-
-        result = _sigtools._order_filterND(a, domain, rank)
-    else:
-        result = ndimage.rank_filter(a, rank, footprint=domain, mode='constant')
+    if a.dtype in [object, np.longdouble]:
+        raise ValueError(f"dtype={a.dtype} is not supported by order_filter")
```
Comment (h-vetinari) [inline scipy/signal/_signaltools.py:1500]: Can we use the same (kind) of check here as for `medfilt`? I.e. something like
```suggestion
    if not (np.issubdtype(volume.dtype, np.integer) 
            or volume.dtype in [np.float32, np.float64]):
        raise ValueError(f"dtype={volume.dtype} is not supported by order_filter")
```

File: scipy/signal/tests/test_signaltools.py [hunk 10]
```diff
@@ -1093,19 +1093,13 @@ def test_types(self, dtype):
...
-            assert_equal(signal.medfilt2d(in_typed).dtype, dtype)
-
-
     @pytest.mark.parametrize('dtype', [np.bool_, np.complex64, np.complex128,
-                                       np.clongdouble, np.float16,])
+                                       np.clongdouble, np.float16, np.object_,
+                                       "float96", "float128"])
     def test_invalid_dtypes(self, dtype):
+        if np.finfo(np.longdouble).dtype != dtype and dtype in ["float96", "float128"]:
+            pytest.skip(f"Platform does not support {dtype}")
```
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1101]: I find the ordering of the conditions hard to follow in terms of the logic behind. Perhaps a comment would help. In particular, the equality between a proper dtype and a sometimes-dtype-sometimes-string is weird, especially as it seems redundant with the `in` check afterwards.

If this was based on my mock [example](https://github.com/scipy/scipy/pull/19673#discussion_r1451617209), then I'd now write it as
```suggestion
        if dtype in ["float96", "float128"] and np.finfo(np.longdouble).dtype.itemsize < 12:
            pytest.skip(f"Platform does not support {dtype}")
```

Comment (j-bowhay) [reply scipy/signal/tests/test_signaltools.py:1101]: I will add a comment explaining the logic but unfortunately the suggestion won't work as a platform will only have float96 or float128 so we still need to skip one of them even if `np.finfo(np.longdouble).dtype.itemsize > 12`

File: scipy/signal/tests/test_signaltools.py [hunk 11]
```diff
@@ -1093,19 +1093,17 @@ def test_types(self, dtype):
...
-                                       np.clongdouble, np.float16,])
+                                       np.clongdouble, np.float16, np.object_,
+                                       "float96", "float128"])
     def test_invalid_dtypes(self, dtype):
+        # Only want to run the test for `float96` and `float128` if the platform has
+        # them. We check what the largest dtype is and compare to see if it is the
+        # requested `float96`, `float128`, or niether.
+        if (dtype in ["float96", "float128"]
+            and np.finfo(np.longdouble).dtype != dtype):
+            pytest.skip(f"Platform does not support {dtype}")
```
Comment (h-vetinari) [inline scipy/signal/tests/test_signaltools.py:1105]: > unfortunately the suggestion won't work as a platform will only have float96 or float128 so we still need to skip one of them even if `np.finfo(np.longdouble).dtype.itemsize > 12`

I still don't understand what the second half here is supposed to achieve. I'm guessing the dtype `==` is overloaded to return `True` if the string matches the underlying dtype.

I also don't understand your point about why it would fail for one or the other if we use my suggestion based on `.itemsize`. Since we're now skipping unconditionally if the dtype is not in `float32` / `float64`, any `longdouble` that's at least 12 bytes long should be rejected.

Comment (j-bowhay) [reply scipy/signal/tests/test_signaltools.py:1105]: > > unfortunately the suggestion won't work as a platform will only have float96 or float128 so we still need to skip one of them even if `np.finfo(np.longdouble).dtype.itemsize > 12`
> 
> I still don't understand what the second half here is supposed to achieve. I'm guessing the dtype `==` is overloaded to return `True` if the string matches the underlying type.

Yes so for example on my platform:
```
In [2]: np.finfo(np.longdouble).dtype == "float96"
Out[2]: False

In [3]: np.finfo(np.longdouble).dtype == "float128"
Out[3]: True
```
so the test will run for `float128` and be skipped for `float96` as desired.

> I also don't understand your point about why it would fail for one or the other if we use my suggestion based on `.itemsize`. Since we're now skipping unconditionally if the dtype is not in `float32` / `float64`, any `longdouble` that's at least 12 bytes long should be rejected.

I am not sure follow as the other condition is `dtype in ["float96", "float128"]` not `float32` / `float64`

```
In [6]: np.finfo(np.longdouble).dtype.itemsize < 12
Out[6]: False

In [7]: np.float96
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[7], line 1
----> 1 np.float96

File ~/miniconda3/envs/scipy-dev/lib/python3.12/site-packages/numpy/__init__.py:333, in __getattr__(attr)
    330     "Removed in NumPy 1.25.0"
    331     raise RuntimeError("Tester was removed in NumPy 1.25.")
--> 333 raise AttributeError("module {!r} has no attribute "
    334                      "{!r}".format(__name__, attr))

AttributeError: module 'numpy' has no attribute 'float96'
```

Comment (h-vetinari) [reply scipy/signal/tests/test_signaltools.py:1105]: OK, I get it now. Here's my suggestion
```suggestion
        # We can only test this on platforms that support a native type of float96 or
        # float128; comparing to np.longdouble allows us to filter out non-native types
        if (dtype in ["float96", "float128"]
                and np.finfo(np.longdouble).dtype != dtype):
            pytest.skip(f"Platform does not support {dtype}")
```

Comment (j-bowhay) [reply scipy/signal/tests/test_signaltools.py]: Thanks, have committed your sugestion

**Review by h-vetinari — APPROVED** [review]
Thank you!

**Comment by h-vetinari** [comment]
@j-bowhay, do you want to keep some of the commit granularity (would probably need a clean-up), or are you fine with a squash-merge?

**Comment by j-bowhay** [comment]
@h-vetinari happy to squash merge
